src/core/sock/sockinfo_tcp.cpp

```diff
@@ -385,6 +396,127 @@ sockinfo_tcp::sockinfo_tcp(int fd, int domain)
     si_tcp_logfunc("done");
 }
 
+void sockinfo_tcp::rx_add_ring_cb(ring *p_ring)
+{
+    if (m_p_group) {
+        m_p_group->add_ring(p_ring);
+    }
+    sockinfo::rx_add_ring_cb(p_ring);
+}
+
+void sockinfo_tcp::set_xlio_socket(const struct xlio_socket_attr *attr)
+{
+    m_xlio_socket_userdata = attr->userdata_sq;
+    m_p_group = reinterpret_cast<xlio_group *>(attr->group);
+
+    bool current_locks = m_ring_alloc_log_rx.get_use_locks();
+
+    m_ring_alloc_log_rx.set_ring_alloc_logic(RING_LOGIC_PER_USER_ID);
+    m_ring_alloc_log_rx.set_user_id_key(reinterpret_cast<uint64_t>(m_p_group));
+    m_ring_alloc_log_rx.set_use_locks(current_locks ||
+                                      (m_p_group->get_flags() & XLIO_GROUP_FLAG_SAFE));
+    m_ring_alloc_logic_rx = ring_allocation_logic_rx(get_fd(), m_ring_alloc_log_rx);
+
+    m_ring_alloc_log_tx.set_ring_alloc_logic(RING_LOGIC_PER_USER_ID);
+    m_ring_alloc_log_tx.set_user_id_key(reinterpret_cast<uint64_t>(m_p_group));
+    m_ring_alloc_log_tx.set_use_locks(current_locks ||
+                                      (m_p_group->get_flags() & XLIO_GROUP_FLAG_SAFE));
+
+    if (!current_locks && (m_p_group->get_flags() & XLIO_GROUP_FLAG_SAFE)) {
+        m_tcp_con_lock = multilock::create_new_lock(MULTILOCK_RECURSIVE, "tcp_con");
+    }
+
+    tcp_recv(&m_pcb, sockinfo_tcp::rx_lwip_cb_xlio_socket);
+    tcp_err(&m_pcb, sockinfo_tcp::err_lwip_cb_xlio_socket);
+    set_blocking(false);
+}
+
+void sockinfo_tcp::add_tx_ring_to_group()
+{
+    ring *rng = get_tx_ring();
+    if (m_p_group && rng) {
+        m_p_group->add_ring(rng);
+    }
+}
+
+void sockinfo_tcp::xlio_socket_event(int event, int value)
+{
+    if (is_xlio_socket()) {
+        /* xlio_group::m_socket_event_cb must be always set. */
+        m_p_group->m_socket_event_cb(reinterpret_cast<xlio_socket_t>(this), m_xlio_socket_userdata,
+                                     event, value);
+    }
+}
+
+/*static*/
+err_t sockinfo_tcp::rx_lwip_cb_xlio_socket(void *arg, struct tcp_pcb *pcb, struct pbuf *p,
+                                           err_t err)
+{
+    sockinfo_tcp *conn = (sockinfo_tcp *)arg;
+
+    NOT_IN_USE(pcb);
+    assert((uintptr_t)pcb->my_container == (uintptr_t)arg);
+
+    // if is FIN
+    if (unlikely(!p)) {
+        return conn->handle_fin(pcb, err);
+    }
+
+    if (unlikely(err != ERR_OK)) {
+        conn->handle_rx_lwip_cb_error(p);
+        return err;
+    }
+
+    tcp_recved(pcb, p->tot_len);
+
+    if (conn->m_p_group->m_socket_rx_cb) {
+        struct pbuf *ptmp = p;
+        while (ptmp) {
+            /* TODO Pass mem_buf_desc_t field intead of pbuf itself as xlio_buf */
+            conn->m_p_group->m_socket_rx_cb(reinterpret_cast<xlio_socket_t>(conn),
+                                            conn->m_xlio_socket_userdata, ptmp->payload, ptmp->len,
```

Comment:
Need to verify that in case of GRO/out-of-order payload and len are always correct for all parts individually.

Reply:
The pbuf chain must be valid (tot_len, len, payload). Otherwise, this would be a bug and RX operations would lead to unexpected behavior. `rx_lwip_process_chained_pbuf()` can be a reference that we rely on the chain validity on each RX operation.